File: hash_table/1_Two_Sum/Solution.js
/**
 * @param {number[]} nums
 * @param {number} target
 * @return {number[]}
 */

var twoSum = function(nums, target) {
    var sums = [];
    var hashTable = {};

    for (var i = 0; i < nums.length; i++) {
        var diff = target - nums[i];

        if (hashTable[diff] !== undefined) {
            sums.push([hashTable[diff], i]);
            return sums;
        }
        // console.log('sums ', sums);

        hashTable[nums[i]] = i; // add the current number to the hash table
        // console.log('hashTable1 ', hashTable);
    }

    // return all pairs of integers that sum to target
};

// console.log(twoSum([1, 2, 3], 3)); // [0,1]
console.log(twoSum([1, 2, 2, 3, 4, 5], 4)); // [0,3]
// console.log(twoSum([1, 2, 4], 7)); // [0,0]
// console.log(twoSum([2, 5, 5, 11], 10)); // [1,2]

File: hash_table/1_Two_Sum/Solution.java
import java.util.Arrays;
import java.util.HashMap;
import java.util.Map;
class Solution {
    public static void main(String[] args) {
        int[] myArr = {1,2,3};
        int[] res = twoSum1(myArr,3);
        System.out.println(Arrays.toString(res));
    }

    // Solution Approach: Two-pass Hash Table
    // public static int[] twoSum(int[] nums, int target) {
    //     Map<Integer, Integer> map = new HashMap<>();
    //     for (int i=0; i< nums.length; i++){
    //         map.put(nums[i], i);
    //     }

    //     for(int i=0; i<nums.length; i++){
    //         int complement = target - nums[i];
    //         if(map.containsKey(complement) && map.get(complement) != i){
    //             return new int[] {i,map.get(complement)};
    //         }
    //     }
    //     throw new IllegalArgumentException("No two sum solution");
    // }

    // Solution Approach: One-pass Hash Table
    public static int[] twoSum1(int[] nums, int target){
        Map<Integer, Integer> map = new HashMap<>();
        for(int i=0; i< nums.length; i++){
            int dif = target - nums[i];
            if(map.containsKey(dif)){
                return new int[] {i, map.get(dif)};
            }
            map.put(nums[i],i);
        }
        throw new IllegalArgumentException("No two sum solution");

    }
}

// To compile:  javac Solution.java 
// To run:      java Solution 